upb/def.c

```diff
@@ -2030,13 +2049,148 @@ static void create_oneofdef(
 
 static str_t *newstr(symtab_addctx *ctx, const char *data, size_t len) {
   str_t *ret = symtab_alloc(ctx, sizeof(*ret) + len);
-  if (!ret) return NULL;
+  CHK_OOM(ret);
   ret->len = len;
   if (len) memcpy(ret->str, data, len);
   ret->str[len] = '\0';
   return ret;
 }
 
+static bool upb_symtab_TryGetChar(const char** src, const char* end, char* ch) {
+  const char* ptr = *src + 1;
+  if (ptr == end) return false;
+  *ch = *ptr;
+  *src = ptr;
+  return true;
+}
+
+static char upb_symtab_TryGetHexDigit(symtab_addctx* ctx, const upb_fielddef* f,
+                                      const char** src, const char* end) {
+  char ch;
+  if (!upb_symtab_TryGetChar(src, end, &ch)) return -1;
+  if (ch >= '0' && ch <= '9') {
+    return ch - '0';
+  }
+  ch = upb_ascii_lower(ch);
+  if (ch >= 'a' && ch <= 'f') {
+    return ch - 'a' + 0xa;
+  }
+  *src -= 1;  // Char wasn't actually a hex digit.
+  return -1;
+}
+
+static char upb_symtab_ParseHexEscape(symtab_addctx* ctx, const upb_fielddef* f,
+                                      const char** src, const char* end) {
+  char hex_digit = upb_symtab_TryGetHexDigit(ctx, f, src, end);
+  if (hex_digit < 0) {
+    symtab_errf(ctx,
+                "\\x cannot be followed by non-hex digit in field '%s' default",
+                upb_fielddef_fullname(f));
+    return 0;
+  }
+  unsigned int ret = hex_digit;
+  while ((hex_digit = upb_symtab_TryGetHexDigit(ctx, f, src, end)) >= 0) {
+    ret = (ret << 4) | hex_digit;
+  }
+  if (ret > 0xff) {
+    symtab_errf(ctx, "Value of hex escape in field %s exceeds 8 bits",
+                upb_fielddef_fullname(f));
+    return 0;
+  }
+  return ret;
+}
+
+char upb_symtab_TryGetOctalDigit(const char** src, const char* end) {
+  char ch;
+  if (!upb_symtab_TryGetChar(src, end, &ch)) return -1;
+  if (ch >= '0' && ch <= '7') {
+    return ch - '0';
+  }
+  *src -= 1;  // Char wasn't actually a octal digit.
+  return -1;
+}
+
+static char upb_symtab_ParseOctalEscape(symtab_addctx* ctx,
+                                        const upb_fielddef* f, char first,
```

Comment:
rather than pass first in, I think it is better to back up the caller by one character and do the parse of all three in here with a simple loop

Reply:
Done.